### PR TITLE
nanvm-lib: Object generalization completes `.`/`[]` (Stage 3)

### PR DESCRIPTION
**Pull request by sasha-gil**
## Summary

Stage 3 (final) of the `.`/`[]` member-access plan. Stacks on #2067 (Stage 2), which stacks on #2066 (Stage 1) — this diff is the incremental change on top of Stage 2.

- `Object::member_access` (`vm/object/member_access.rs`, new, `pub(crate)`): a `Number` key is stringified first — `ToString(number)`, via a new `number_to_string` helper extracted from `StringCoercion::number` so both share the same rules rather than duplicating them — since a plain object doesn't special-case a numeric-looking key the way `Array`/`String` do (`{0:'a'}[0]` and `{0:'a'}['0']` must agree). A `String` key goes straight to the existing `own_property`, unchanged.
- `own_property` stays a separate method, not a thin wrapper: `own` is an `Op2Id` whose key operand is a full `exp`, and a non-`String` key reaching it is a bug upstream — reported as a hard `Err`. `.`/`[]`'s `index` is schema-restricted to `number | string`, and both are legitimate keys — `Object::member_access` only answers `None` for anything else, never errors. Conflating the two would weaken `own`'s stricter contract.
- `Any::member_access` now dispatches every receiver: `Array`, `String`, `Object` to their own logic, `Number`/`Boolean`/`BigInt`/`Function` straight to `undefined` (no own properties). The `todo!()` is gone — this completes the operator.
- `nanvm-lib/README.md`'s `.`/`[]` row flips to `[x]`.

### Review round: two fixes from Copilot's review

- **`number_to_string` now matches ECMAScript `Number::toString()`.** It previously used Rust's `f64::to_string()`, which never switches to exponential notation — `(1e21).to_string()` was a 22-digit integer where JS gives `"1e+21"`, and `(1e-7).to_string()` was `"0.0000001"` where JS gives `"1e-7"`. This was latent in `StringCoercion::number` before this PR ever touched it, but `Object::member_access` gives it a second, newly-reachable path, and it's code this PR touches either way. Replaced with the actual ECMA-262 notation rule applied to the digit sequence Rust's own `{:e}` formatting computes, verified against a 25-case standalone probe (every notation boundary, `Number.MAX_VALUE`, `Number.MIN_VALUE`) plus 4 new dedicated unit tests.
- **The plan doc is deleted, not kept.** `AGENTS.md`/`CONTRIBUTING.md` both say to delete a `todo/` issue file once the PR fixing it lands. The earlier "kept rather than deleted" call (modeled on `fjs/effects/todo/io-effect-migration.md`) didn't actually hold up on a closer look: that precedent is cited by genuinely independent, pre-existing documents, where this one was only cited by its own PR's new code and a README line that's just as easy to write inline. Deleted `nanvm-lib/todo/member-access-operator.md`, folding its two real citations (the README row, three `member_access` doc comments) into self-contained wording.

## Test plan

- [x] 6 new unit tests for `Object::member_access` (string key, numeric key stringification, `-0` parity, no-match, unrelated key type) plus 2 new dispatcher-level tests in `Any::member_access`
- [x] 4 new unit tests for `number_to_string` (exponential boundaries, plain notation, extreme magnitudes, non-finite/NaN)
- [x] `cargo test -p nanvm-lib` — 219 passed (full suite, incl. doctests), up from 208 in Stage 2
- [x] `cargo fmt -p nanvm-lib -- --check` — clean
- [x] `cargo clippy -p nanvm-lib --lib` — clean
- [x] `number_to_string`'s notation-boundary fix independently verified via a standalone Rust probe against known `Number.prototype.toString()` outputs before integrating
- [x] Confirmed via repo-wide grep that no file references the deleted plan doc after the edits

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Xjs4wQV6tLvf9fec4F4pom

### CODE REVIEW

File: nanvm-lib/src/vm/string_coercion.rs [hunk 1]
```diff
@@ -80,3 +75,18 @@ impl<A: IVm> Dispatch<A> for StringCoercion {
...
+/// around it) — also reused directly by `Object::member_access`
+/// (`vm/object/member_access.rs`) to stringify a numeric key before a
+/// plain-object lookup, the same way real JS's `ToPropertyKey` runs ahead
+/// of `[[Get]]`.
+pub(crate) fn number_to_string<A: IVm>(v: f64) -> String<A> {
+    match v {
+        f64::INFINITY => "Infinity".into(),
+        f64::NEG_INFINITY => "-Infinity".into(),
+        -0.0 => "0".into(),
+        v => v.to_string().as_str().into(),
```
[the remark this answers is not in the document]
Comment (sasha-gil) [reply nanvm-lib/src/vm/string_coercion.rs]: Confirmed and fixed in f3798c1 — verified the exact boundaries you named (`1e21`/`1e-7`) plus `1e20`/`1e-6` (the adjacent non-exponential cases), `Number.MAX_VALUE`/`Number.MIN_VALUE`, and negatives, against a standalone probe before integrating. `number_to_string` now applies ECMA-262's actual notation rule to the digit sequence Rust's `{:e}` formatting computes, with 4 dedicated unit tests covering the boundaries.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by o2alexanderfedin — COMMENTED** [review]
Reviewed as a stacked PR: delta measured against the base branch tip `64eb554ba7d7d9b99472619b78e7d2c58f8f0626` (`claude/nanvm-member-access-stage2`, also the merge-base; not yet an ancestor of `origin/main`).

All four CI gates green locally from `nanvm-lib/` (exit 0 each): `cargo test` (219 lib + 1 integration + 5 doc), `cargo test --release`, `cargo clippy -- -D warnings`, `cargo clippy --release -- -D warnings`. The i686 cross jobs need a toolchain this machine lacks — not run. `gh pr checks 2068`: all 14 pass.

**One finding — `number_to_string` diverges from JS on exact round-half-to-even ties.**

I ran a differential of `number_to_string` against `node`'s `String(v)` over ~60k f64 bit patterns (named boundary values plus random patterns and random scaled decimals), comparing by bit pattern. 3 distinct values disagree:

| bits | `node` / CPython | this PR |
|---|---|---|
| `c23a0480a70a2400` | `-111744689930.14062` | `-111744689930.14063` |
| `c24e2a807a3c5a00` | `-259124163704.70312` | `-259124163704.70313` |
| `c24ed800216b1200` | `-264945812182.14062` | `-264945812182.14063` |

The first is exactly `-111744689930.140625`. Both 17-digit candidates round-trip (checked: `"…14062".parse::<f64>()` and `"…14063".parse::<f64>()` both equal the value), so they are equidistant from *m*, and <https://tc39.es/ecma262/#sec-numeric-types-number-tostring> step 5 then requires the **even** `s` — what V8 emits. Rust's shortest-repr rounds the tie away from zero, and `{}` gives the same as `{:e}`, so this is not an artifact of the `{:e}` choice. Since Stage 3 routes object keys through this helper, `({'-111744689930.14062': 1})[-111744689930.140625]` would read `undefined` here but `1` in JS. Rare, but the helper's doc comment claims spec conformance.

Everything else checked out: object key coercion for `-0`/`NaN`/`±Infinity`/`1e21`/`1e-7`/`0.1+0.2` matches JS; no panic reachable from a caller-supplied key across `Array`/`String`/`Object` receivers over an adversarial key corpus. Positive controls: deleting `string_to_index`'s `len == 0` guard panics at `member_access.rs:45`; swapping `number_to_string` for `f64::to_string` fails the coercion probe — both restored and re-green. The deleted plan doc's Stage 3 items (Object dispatch, `undefined` for `Number`/`Boolean`/`BigInt`/`Function`, README row `[x]`) are all done at this head, and nothing in the tree still references `member-access-operator` (grep exit 1, positive-controlled).

**Review by sergey-shandar — APPROVED** [review]

**Review by sergey-shandar — CHANGES_REQUESTED** [review]
Need to address number conversion.

**Comment by sasha-gil** [comment]
@sergey-shandar Pushed a fix for the number conversion issue: `number_to_string`'s `Number::toString` now applies ECMA-262's round-half-to-even tie-break exactly, via `BigInt<A>` arbitrary-precision comparison (already `nanvm-lib`'s own facility, so no new dependency), rather than relying solely on Rust's `{:e}` shortest-round-trip formatter, which can break an exact tie the other way.

- Added `round_tie_to_even`/`mantissa_exp2` in `string_coercion.rs`, wired into `js_digits_to_string`.
- New tests cover the three bit patterns the earlier review found (`0xc23a0480a70a2400`, `0xc24e2a807a3c5a00`, `0xc24ed800216b1200`), plus regression guards on `f64::MAX`/`f64::MIN_POSITIVE`/`5e-324` (an earlier, rejected fix attempt using a round-trip heuristic instead of exact arithmetic had broken those).
- `cargo test -p nanvm-lib` — 221 passed (up from 219), `cargo clippy -p nanvm-lib --lib -- -D warnings` and `cargo fmt -p nanvm-lib -- --check` both clean.
- Deleted `nanvm-lib/todo/number-to-string-tie-breaking.md`, which tracked this as a deferred follow-up — now fixed directly instead.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sergey-shandar** [comment]
@codex review

**Review by o2alexanderfedin — APPROVED** [review]
Delta review since `f3798c1d`; measured against base `4e2f9776cc9ae484e6c0941eb77008b9a02b8d8d` (merge-base with `main`, now that #2066/#2067 have landed). The PR's own new content over that marker is exactly `nanvm-lib/README.md` + `src/vm/string_coercion.rs` — the Stage-3 `member_access` blobs are byte-identical, and `b0ffe220` is a plain merge of `main`.

**The tie-break finding is fixed.** I re-ran the differential against `node`'s `String(v)` over **649,923 f64 bit patterns** (249,890 uniform-random `u64`, plus 400,033 with the exponent field banded to ±60 around 1023 — where shortest-repr ties actually live), including the three values I named and the ordinary/boundary set (0, −0, ±1, 0.1, 0.1+0.2, 1/3, 1e20, 1e21, 1e−6, 1e−7, 5e−324, MAX, MIN_POSITIVE, 2^53±1, EPSILON, 1e±300, 2^−24): **0 disagreements**.

That zero is controlled two ways. Perturbing one emitted string made the harness report exactly 1 mismatch. And short-circuiting `round_tie_to_even` to `return s` turned the same banded corpus into **1,640 disagreements** — so the corpus really does exercise the fix, and the fix closes all of them. Mutating away the `is_legitimate_candidate` guard from `f5296ef2` turns `power_of_two_boundary_is_not_a_tie` red; both mutations restored and re-green.

The arithmetic holds up on reading too: `s ≥ 1` so `2 * s - 1` cannot underflow, 17 digits keeps `2 * s + 1` and `2 * mantissa` far inside `u64`, and both `pow2`/`pow10` arguments are non-negative by construction, so every `expect` is unreachable. A corrected candidate with a trailing zero would be a shorter round-tripping spelling Rust would already have emitted — the round-trip guard rules it out.

`7c122e71`'s followup doc is deleted again by `5fe4b94e`, with no dangling references left.

Gates from `nanvm-lib/`, all exit 0: `cargo test`, `cargo test --release`, `cargo clippy -- -D warnings`, `cargo clippy --release -- -D warnings`. JS suite 7094/0. The i686 cross jobs need a toolchain this machine lacks — not run here; CI is green on all 19 checks.

**Comment by sergey-shandar** [comment]
@codex review

File: nanvm-lib/src/vm/string_coercion.rs [hunk 2]
```diff
@@ -80,3 +77,297 @@ impl<A: IVm> Dispatch<A> for StringCoercion {
...
+    let digits: std::string::String = mantissa.chars().filter(|&c| c != '.').collect();
+    let k = digits.len() as i64;
+    let n = exp.parse::<i64>().expect("`{:e}`'s exponent is an integer") + 1;
+
+    let s = digits
+        .parse::<u64>()
+        .expect("`digits` is a run of ASCII decimal characters");
+    let m = n - k;
+    let (mant, exp2) = mantissa_exp2(v);
+    let corrected = round_tie_to_even::<A>(v, mant, exp2, s, m);
```
[the remark this answers is not in the document]
Comment (sasha-gil) [reply nanvm-lib/src/vm/string_coercion.rs:133]: Confirmed and fixed in 4481e1b. Reordered to two cheap filters first — a correction only ever moves an *odd* digit to an *even* neighbor that *round-trips* (plain `str::parse`, no `BigInt`) — both conditions the correctness fix already required, just checked before the arithmetic that would otherwise re-derive them. `BigInt` is now only ever constructed for the rare case that clears both.

Re-benchmarked release-mode over 100k random finite values: ~126ms, down from the ~467ms you measured (vs. ~22ms for Rust's own `{:e}` alone — the remainder is the unavoidable round-trip `parse` cost for odd-digit values, not `BigInt`). Confirmed behavior-preserving: byte-identical output to the prior version across the same ~14k power-of-two-boundary values and 80k random bit patterns, all still matching Node.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by sasha-gil** [comment]
Pushed a performance fix (4481e1b) for Codex's finding on the round-trip guard: the exact `BigInt` tie-break was running unconditionally on every `number_to_string` call, not just numeric object keys. Reordered so two cheap checks (already required for correctness — see the doc comment) run first: a correction only ever moves an odd digit to an even neighbor that round-trips, checked via plain `str::parse` before any `BigInt` is constructed. `BigInt` now only runs for the rare case that clears both filters.

- Release-mode benchmark: ~126ms/100k random finite values, down from ~467ms.
- Verified behavior-preserving: byte-identical output to the prior version across the same ~14k power-of-two-boundary values and 80k random f64 bit patterns (all still matching Node).
- `cargo test -p nanvm-lib` — 222 passed, `cargo clippy -p nanvm-lib --lib -- -D warnings` and `cargo fmt -p nanvm-lib -- --check` both clean.

@sergey-shandar this should address both the correctness finding and the performance one from Codex's review.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by sergey-shandar — APPROVED** [review]
